### PR TITLE
feat(mcp): complete end-to-end protocol 2026-07-28 support

### PR DESCRIPTION
**Pull request by Laisky**
## Summary

Complete one-api's end-to-end MCP gateway implementation for protocol `2026-07-28` across both roles:

- **downstream server:** one-api exposes an authenticated aggregate MCP endpoint;
- **upstream client:** one-api connects to configured MCP servers;
- **gateway middle layer:** synchronized tool descriptors remain lossless and case-sensitive through persistence, policy, routing, and re-exposure.

The implementation retains bounded compatibility with initialization/session protocol revisions `2025-11-25`, `2025-06-18`, and `2025-03-26`, without weakening the stateless `2026-07-28` request model.

### MCP server side

- serves `server/discover`, paginated `tools/list`, and `tools/call` on `/mcp`;
- validates per-request `_meta`, `Mcp-Protocol-Version`, `Mcp-Method`, `Mcp-Name`, and schema-derived `Mcp-Param-*` headers;
- returns deterministic, private-cacheable tool pages with opaque catalog-fingerprint cursors;
- preserves modern `resultType`, `structuredContent`, `isError`, `inputRequests`, `requestState`, `_meta`, and extension fields;
- delegates recognized legacy traffic to the initialization/session handler;
- bounds both modern and legacy request bodies and redacts internal failures.

### MCP client side

- uses stateless `2026-07-28` requests first for discovery, tool listing, and tool calls;
- follows every `tools/list` cursor with loop and page limits;
- correlates JSON and SSE responses by the exact JSON-RPC request ID;
- bounds remote response bodies;
- sends normalized empty argument objects for zero-argument tools;
- derives and validates `x-mcp-header` parameter headers;
- preserves multi-round-trip `inputResponses` and `requestState`;
- falls back only for failures that genuinely indicate a legacy endpoint;
- serializes shared negotiated-header state safely for concurrent fallback;
- prevents credentialed network endpoints from using plaintext HTTP and rejects unsafe credential redirects.

### Aggregate catalog and persistence

- stores the complete normalized upstream descriptor, including current and unknown extension fields;
- retains searchable columns for administration without treating them as the wire-format source of truth;
- preserves exact case-sensitive tool names;
- atomically replaces one server's catalog and rolls back on failure;
- preserves local UUID, status, pricing, and creation metadata for unchanged exact names;
- re-emits the complete descriptor after policy filtering and server-name qualification;
- rejects malformed, stale, or out-of-range aggregate pagination cursors.

## Compatibility boundary

This PR implements the MCP `tools` capability that one-api advertises. It deliberately does **not** invent a non-standard MCP `tools/search` method.

OpenAI Responses API `tool_search`, `defer_loading`, and `tool_search_call` are model-orchestration features above MCP and remain a separate integration concern. The deterministic aggregate catalog introduced here is suitable as their backing index.

## Behavioral coverage

Tests cover:

- modern client-to-server contract behavior;
- complete descriptor and unknown extension-field round trips;
- case-sensitive tool coexistence and exact invocation;
- atomic catalog replacement and local metadata preservation;
- upstream and downstream pagination, malformed and stale cursors;
- modern/legacy protocol routing and omitted legacy initialize parameters;
- fallback across `2025-11-25`, `2025-06-18`, and `2025-03-26`;
- zero-argument calls, parameter-header encoding, numeric equivalence, and mismatch rejection;
- JSON/SSE response-ID correlation and bounded response bodies;
- multi-round-trip interim results without duplicate billing;
- optional JSON `null` compatibility;
- credentialed transport and redirect protections;
- concurrent modern-to-legacy fallback.

## Validation

Verified against the current PR head from a clean clone:

- `git diff --check`
- `go test -count=1 ./relay/mcp ./model ./controller`
- `go run ./tools/analyzers/noentityresponse/cmd/noentityresponse ./...`
- `go vet ./...`
- `go test -race ./...`
- `make build-frontend-modern`

The normal GitHub Actions `lint` and `pr-check` workflows also completed successfully for the current PR head.

### CODE REVIEW

File: controller/mcp_proxy_latest.go [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+	return e.Err.Error()
+}
+
+// isModernMCPRequest reports whether a request uses the MCP 2026-07-28 request model.
+func isModernMCPRequest(c *gin.Context, req mcpRPCRequest) bool {
+	if strings.TrimSpace(req.Method) == "server/discover" {
+		return true
+	}
+	if strings.TrimSpace(c.GetHeader(mcp.ProtocolVersionHeader)) != "" {
+		return true
```
[the remark this answers is not in the document]
Comment (Laisky) [reply controller/mcp_proxy_latest.go]: Addressed. The current dispatcher classifies `Mcp-Protocol-Version` using `mcp.IsLegacyProtocolVersion`; supported legacy versions are delegated to `MCPProxy` rather than treated as modern merely because the header is present. The controller regression suite covers legacy initialize and request routing through the shared `/mcp` endpoint.

File: controller/mcp_proxy_latest.go [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+	"github.com/gin-gonic/gin"
+
+	"github.com/Laisky/one-api/common/config"
+	"github.com/Laisky/one-api/model"
+	"github.com/Laisky/one-api/relay/mcp"
+)
+
+const modernMCPMaxRequestBytes = 4 << 20
+
+// MCPProxyLatest dispatches MCP 2026-07-28 requests while preserving the legacy endpoint behavior.
```
[the remark this answers is not in the document]
Comment (Laisky) [reply controller/mcp_proxy_latest.go]: Addressed. The added handlers, client methods, helpers, and regression tests now use complete GoDoc blocks that start with the symbol name and document purpose, parameters, and return behavior, including no-return handlers.

File: controller/mcp_proxy_latest.go [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+
+	body, err := io.ReadAll(io.LimitReader(c.Request.Body, modernMCPMaxRequestBytes+1))
+	if err != nil {
+		respondMCPModernError(c, nil, http.StatusBadRequest, mcpErrParseError, errors.Wrap(err, "read mcp request"), nil)
+		return
+	}
+	c.Request.Body = io.NopCloser(bytes.NewReader(body))
+	if len(body) > modernMCPMaxRequestBytes {
+		respondMCPModernError(c, nil, http.StatusRequestEntityTooLarge, mcpErrInvalidRequest, errors.New("mcp request body exceeds 4 MiB"), nil)
+		return
```
[the remark this answers is not in the document]
Comment (Laisky) [reply controller/mcp_proxy_latest.go]: Addressed. The handler now classifies legacy protocol headers before modern validation and applies separate bounded limits: 32 MiB for delegated legacy traffic and 4 MiB for modern traffic. The reconstructed body is bounded before either handler decodes it, preserving legacy compatibility without an unbounded allocation path.

File: relay/mcp/client_latest.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+		return nil, errors.Wrap(err, "normalize mcp tool arguments")
+	}
+	parameterHeaders, err := ToolArgumentHeaders(tool.InputSchema, argumentMap)
+	if err != nil {
+		return nil, errors.Wrapf(err, "derive mcp headers for tool %q", name)
+	}
+	params := map[string]any{
+		"name":      name,
+		"arguments": arguments,
+	}
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/mcp/client_latest.go:160]: Addressed in `53acb94`. `CallToolLatestWithOptions` now serializes the normalized `argumentMap`, so omitted or nil arguments are sent as `{}` rather than `null`. `TestCallToolLatestNormalizesNilArguments` verifies the actual wire payload against a strict mock upstream.

File: relay/mcp/client_latest.go [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+		}
+		if json.Unmarshal(candidate, &envelope) != nil {
+			continue
+		}
+		if id, ok := envelope.ID.(string); ok && id == requestID {
+			return candidate, nil
+		}
+	}
+	if len(fallback) != 0 {
+		return fallback, nil
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/mcp/client_latest.go]: Addressed. JSON and SSE responses now pass through strict request-ID correlation. `extractMCPResponseEnvelope` returns an error when no SSE event matches the active request and never falls back to the last event. The regression suite includes unrelated-ID SSE and mismatched JSON response cases.

File: relay/mcp/headers.go [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+	}
+	return strconv.FormatInt(signed, 10), nil
+}
+
+// headerValueRequiresEncoding reports whether a value requires the MCP Base64 sentinel.
+func headerValueRequiresEncoding(value string) bool {
+	if !utf8.ValidString(value) || strings.TrimSpace(value) != value {
+		return true
+	}
+	if strings.HasPrefix(value, mcpBase64SentinelPrefix) && strings.HasSuffix(value, mcpBase64SentinelSuffix) {
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/mcp/headers.go]: Addressed. `headerValueRequiresEncoding` now requires Base64 sentinel encoding when either the prefix or suffix boundary is present. Regression coverage includes one-sided sentinel-like values so an encoded tool name or parameter round-trips instead of being rejected by the decoder.

File: relay/mcp/types.go [hunk 7]
```diff
@@ -22,55 +27,139 @@ func (t *ToolDescriptor) UnmarshalJSON(data []byte) error {
...
+		if schema, ok := value.(map[string]any); ok {
+			return schema
+		}
+		encoded, err := json.Marshal(value)
+		if err != nil {
+			continue
+		}
+		var schema map[string]any
+		if err := json.Unmarshal(encoded, &schema); err == nil {
+			return schema
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/mcp/types.go]: Addressed. `ToolDescriptor.UnmarshalJSON` now returns the `decodeOptionalObject` error for a present malformed `inputSchema`/legacy alias instead of silently replacing it with nil. Nullable behavior is limited to fields that are optional by protocol; malformed non-object schemas still fail synchronization.

File: controller/mcp_call_latest.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "build mcp tool candidates for %q", toolName)
+	}
+	candidates = filterExactMCPToolCandidates(candidates, toolName)
+	if len(candidates) == 0 {
+		return nil, errors.Errorf("no eligible MCP tool found for exact name %q", toolName)
+	}
+
+	startedAt := time.Now()
```
[the remark this answers is not in the document]
Comment (Laisky) [reply controller/mcp_call_latest.go]: Not applied: this value is not a persisted or API wall-clock timestamp. It is used only as the start value for `helper.CalcElapsedTime`, so retaining `time.Now()` preserves Go's monotonic clock reading for duration measurement. Calling `.UTC()` changes location presentation and strips the monotonic component; there is no timezone-sensitive timestamp emitted from this value.

File: .github/workflows/pr387-source-snapshot.yml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        uses: actions/setup-go@v7.0.0
+        with:
+          go-version-file: go.mod
+
+      - name: Pack validation assets
+        run: |
+          set -euo pipefail
+          tar --exclude=.git -czf /tmp/one-api-pr387-source.tar.gz .
+          tar -C "$(dirname "$(go env GOROOT)")" -czf /tmp/go-1.26.3-linux-x64.tar.gz "$(basename "$(go env GOROOT)")"
+          tar -C "$(go env GOPATH)" -czf /tmp/go-module-cache.tar.gz pkg/mod
```
[the remark this answers is not in the document]
Comment (Laisky) [reply .github/workflows/pr387-source-snapshot.yml]: No production fix is needed because this comment refers to a temporary source-snapshot workflow used only during review recovery. That workflow and all `pr387-*` helper workflows/scripts were removed from the branch before the source result was finalized, so the affected path is no longer part of the PR or its runtime/CI surface.

File: relay/mcp/client.go [hunk 10]
```diff
@@ -20,100 +19,122 @@ import (
...
-			merged["X-API-Key"] = server.APIKey
+	if server != nil {
+		switch strings.ToLower(server.AuthType) {
+		case model.MCPAuthTypeBearer:
+			if server.APIKey != "" {
+				merged["Authorization"] = "Bearer " + server.APIKey
+			}
+		case model.MCPAuthTypeAPIKey:
+			if server.APIKey != "" {
+				merged["X-API-Key"] = server.APIKey
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/mcp/client.go:114]: Addressed in `41a8070` with defense in depth and behavioral regressions. `MCPServer.NormalizeAndValidate` rejects credentialed remote `http://` configurations, while `mcpSecurityRoundTripper` repeats the policy immediately before network I/O so older persisted rows or directly constructed clients cannot bypass it. HTTPS is accepted; unauthenticated HTTP remains compatible; credentialed HTTP is limited to loopback hosts; HTTPS downgrade and credentialed cross-origin redirects remain blocked. Coverage is in `model/mcp_server_security_test.go` and `relay/mcp/transport_security_test.go`.

**Comment by Laisky** [comment]
## Developer handoff: current CI blockers and recommended recovery plan

This note is for the next developer taking over PR #387. It supersedes the stale **“all checks passed”** statement currently present in the PR description.

### Current state

- PR head: `c4f6fa9fd299cfcafc91207ef5dad0b8240eb718`
- GitHub Actions tested synthetic merge commit: `620fec8eeb9cccd40388d712b32e92b39a993eb9`
- Base used by that merge: `da318831c6323c354efd4fa8211bf14e2a89aa8b` (`develop`)
- All existing inline review threads are resolved.

Latest checks on this tree:

- ✅ Go Vet
- ✅ goroutine guardrail
- ✅ boundary analyzer
- ✅ PostgreSQL migration tests
- ✅ MySQL migration tests
- ✅ behavior contract tests
- ❌ Go unit tests
- ❌ Go unit tests (PostgreSQL)

The failing tests form three coherent clusters. They appear to be compatibility problems in tests/request construction rather than unrelated failures across the repository.

---

## 1. Fix typed-nil JSON-RPC params being serialized as `null`

### Observed failures

The following tests in `relay/mcp/client_test.go` all fail through the same strict mock-server error:

- `TestStreamableHTTPClient_ListToolsHappyPath`
- `TestStreamableHTTPClient_OmitsNullParamsOnRequests`
- `TestStreamableHTTPClient_CapturesSessionIDAndPropagates`
- `TestStreamableHTTPClient_NotificationsHaveNoID`
- `TestStreamableHTTPClient_HandlesSSEResponse`
- `TestStreamableHTTPClient_AcceptHeader`
- `TestStreamableHTTPClient_AuthBearer`
- `TestStreamableHTTPClient_AuthAPIKey`

The reported error is:

```text
mcp protocol error: status 400: code -32700: Parse error: Invalid JSON-RPC message
```

### Root cause

`ListTools` currently declares:

```go
var params map[string]any
```

When no cursor exists, that value is a nil map. However, it is passed to `doRPC`/`doRPCRaw` through an `any` parameter. An interface containing a typed nil map is **not equal to nil**, so this condition succeeds:

```go
if params != nil {
    payload["params"] = params
}
```

The resulting request contains:

```json
"params": null
```

The strict mock intentionally rejects this. Therefore, do **not** weaken the strict fixture or change it to accept `params: null`.

### Recommended fix

At minimum, ensure `ListTools` passes an actual nil interface when the cursor is empty, for example by declaring `params` as `any` and assigning a map only when a cursor exists:

```go
var params any
if cursor != "" {
    params = map[string]any{"cursor": cursor}
}
```

A more defensive option is to centralize request-payload construction and omit nil-like map/slice/pointer/interface values. If that broader fix is chosen, keep it small and add direct tests because reflection-based nil handling can easily panic on non-nilable kinds.

Add or retain a regression that passes `map[string]any(nil)` through the real request builder and verifies that the encoded JSON has no `params` member at all.

---

## 2. Make controller MCP fixtures echo the actual JSON-RPC request ID

### Observed failures

These controller tests fail because the client expects its generated request ID, while the fake upstream returns the hard-coded ID `"1"`:

- `TestCallMCPToolForUser_FreeToolEmitsLog`
- `TestCallMCPToolForUser_PaidToolDeductsQuotaAndLogs`
- `TestCallMCPToolForUser_ErrorResultIsNotLoggedOrCharged`
- `TestMCPProxy_HTTPHandlerLogsFreeToolCall`

Typical failure:

```text
mcp response id mismatch: expected "<generated UUID>", got "1"
```

### Root cause

`setupMCPProxyTest` correctly decodes the incoming request into `rpc.ID`, and its `initialize` response already echoes that ID. However, the default `respondPayload` and several overrides still construct tool-call responses with:

```json
"id": "1"
```

Strict response correlation was intentionally added in this PR and should remain enabled.

### Recommended fix

Refactor the fixture so the configurable tool responder receives the decoded request ID, for example:

```go
respondPayload func(id any) ([]byte, int)
```

Then use the existing `marshalIDForFixture` helper to echo the ID in every JSON-RPC tool response. Update each override in the affected tests as part of the same change.

An even cleaner fixture shape is to let test overrides supply only the `result` or `error` object while the common HTTP handler always wraps it in a JSON-RPC envelope using `rpc.ID`. Either approach is acceptable; the important contract is that a successful mock server must echo the exact incoming ID, including string versus numeric type.

Do **not** relax `validateMCPResponseID` or reintroduce “accept the last SSE event” behavior. Keep at least one negative test proving that a genuinely mismatched response ID is rejected.

---

## 3. Update stale discovery expectations to include every supported version

### Observed failures

- `TestMCPProxyLatestDiscover`
- `TestMCPProxyLatestClientServerContract`

Current expected value:

```text
["2026-07-28"]
```

Actual value:

```text
["2026-07-28", "2025-11-25", "2025-06-18"]
```

### Root cause and intended contract

The server implementation deliberately returns:

```go
mcp.SupportedProtocolVersions()
```

That helper is documented as returning the modern version followed by the supported legacy compatibility versions. This matches the PR’s compatibility goal and the dispatcher’s support for both legacy protocol versions.

### Recommended fix

Update both tests to compare against:

```go
mcp.SupportedProtocolVersions()
```

Avoid duplicating a hard-coded version list in controller tests. It is also useful to assert separately that the first entry remains `mcp.ProtocolVersion`, because ordering represents preference:

```go
versions := mcp.SupportedProtocolVersions()
require.NotEmpty(t, versions)
require.Equal(t, mcp.ProtocolVersion, versions[0])
```

Only change production discovery back to a single version if there is an explicit product decision that `supportedVersions` means “modern-only”; that would be a protocol-contract change and would require corresponding documentation and compatibility review.

---

## Suggested implementation order

1. Fix typed-nil params and run the focused `relay/mcp` tests.
2. Fix the controller fixture’s JSON-RPC ID echo and run the focused controller tests.
3. Update the two discovery assertions to use `mcp.SupportedProtocolVersions()`.
4. Run both packages under the race detector.
5. Run the complete local CI command.
6. Re-run both GitHub workflows on the exact updated PR head and confirm the generated merge ref is green.

### Focused commands

```bash
go test ./relay/mcp \
  -run 'TestStreamableHTTPClient_(ListToolsHappyPath|OmitsNullParamsOnRequests|CapturesSessionIDAndPropagates|NotificationsHaveNoID|HandlesSSEResponse|AcceptHeader|AuthBearer|AuthAPIKey)$' \
  -count=1

go test ./controller \
  -run 'Test(MCPProxyLatestDiscover|MCPProxyLatestClientServerContract|CallMCPToolForUser_FreeToolEmitsLog|CallMCPToolForUser_PaidToolDeductsQuotaAndLogs|CallMCPToolForUser_ErrorResultIsNotLoggedOrCharged|MCPProxy_HTTPHandlerLogsFreeToolCall)$' \
  -count=1

go test -race ./relay/mcp ./controller -count=1
```

### Full validation

CI resolves the Go version from `go.mod` and currently uses Go `1.26.3`. The full unit-test job also installs `ffmpeg`/`ffprobe`.

```bash
go vet ./...

go test -race -cover -coverprofile=coverage.txt \
  -count=1 -timeout 20m ./...
```

To reproduce the same tree GitHub tests, fetch both PR refs rather than testing only the branch head:

```bash
git fetch -f origin \
  refs/pull/387/head:refs/heads/pr-387-head \
  refs/pull/387/merge:refs/heads/pr-387-merge

git switch pr-387-merge
```

The merge ref changes whenever either the PR head or `develop` changes, so record the tested merge SHA in the final PR update.

---

## Regression guardrails

While repairing CI, preserve these behaviors already implemented or reviewed in this PR:

- strict JSON-RPC response-ID correlation for JSON and SSE;
- omission of absent request params rather than serializing `null`;
- discovery of modern plus explicitly supported legacy versions in preference order;
- legacy request routing for `2025-11-25` and `2025-06-18`;
- normalized zero-argument tool calls using `{}` rather than `null`;
- no quota charge or usage log for `input_required` interim results;
- bounded request and response bodies;
- credentialed remote MCP requests require HTTPS, except loopback development endpoints;
- HTTPS downgrade and credentialed cross-origin redirects remain blocked;
- arbitrary custom headers are currently treated conservatively as potential credentials. Do not loosen that policy merely to fix the failures above; it is not the cause of the current JSON parse errors. Any compatibility change there should be explicit and covered by transport tests.

## Definition of done

This PR is ready for another merge review only when all of the following are true:

- the focused tests above pass;
- `go test -race ./relay/mcp ./controller -count=1` passes;
- the repository-wide race/coverage command passes;
- both `lint` and `pr-check` are green on the same updated head;
- the generated PR merge ref against current `develop` is green;
- the PR description’s validation section is updated with the new head SHA, merge SHA, commands, and results;
- no new unresolved review thread remains.

The first three fixes should be handled together because they are small, tightly related test/contract corrections. Avoid broad protocol rewrites until those known failures are removed and a new CI run exposes any remaining issue.
